src/view/subgraph.rs

```
@@ -321,6 +309,58 @@ where
     }
 }
 
+impl<G: LinkMut> Subgraph<G>
+where
+    G::LinkEndpoint: Into<PortIndex>,
+{
+    /// Copies all the nodes and edges in this subgraph into the parent graph.
+    /// If there are any boundary edges, these will also be copied but keeping
+    /// the same *external* end port (this will fail unless the underlying graph
+    /// is a [MultiView]).
+    ///
+    /// Returns a map from node indices within this subgraph, to the indices
+    ///    of the newly-created nodes in the parent graph (they are not in this subgraph!);
+    /// or a LinkError in which case the underlying graph will not have had any nodes added
+    ///    (however subports may have been for a [MultiView]).
```

It might be cleanest for the API if you would differentiate between the "expected" error of finding a boundary edge in a non-Multiview graph, vs the "seriously wrong error" of a link failing to get added for inscrutable reasons (and in which case it is hard to know whether other ports have already been added).

In the first case you could return a better error than a LinkError... But this might be unnecessarily complicating things.

---

I think that's fair, it did make me think "ugh, hmmm" a bit while I was writing it. Lemme have a go...

---

Hmmm, so the first thing I note is that any `LinkError` for internal edges must be a bug in `copy_in_parent` - perhaps we should `unwrap` that rather than return to the user. Then we can say any LinkError must be from the boundary edges. Moreover, the only possibility is `AlreadyLinked` if we are copying in a plain portgraph - we could panic/unwrap on any other variant as again that must be a bug in `copy_in_parent`.  It would be nice thus to define an error type that for `MultiPortgraph`s is `Infallible` but for plain `Portgraph`s is *only* the "AlreadyLinked" variant of LinkError...

that's not terribly consistent with the rest of portgraph, though.

---

- I would remove the subports-added on failure note. Those are implementation details and not visible to the user unless they peel the multiport layer off.

- Unwrap (or rather `expect`) on internal bugs seems reasonable. Errors should be for _user bugs_.

- Use the standard rustdoc headers to clean this up a bit

```rust
    /// # Returns
    ///
    ///   A map from node indices within this subgraph, to the indices
    ///   of the newly-created nodes in the parent graph (they are not in this subgraph!).
    ///
    /// # Errors
    ///
    /// - [CopySubgraphError::CantCopyBoundary] if the boundary was not empty, but graph doesn't support support multilinks. The underlying graph remains unchanged.
```